### ISSUE TITLE
Modelica.Thermal.HeatTransfer.Sensors in 3.2.1beta1 are not backward compatible

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1125]
**Reported by mwetter on 24 May 2013 22:22 UTC**
Models in Modelica.Thermal.HeatTransfer.Sensors added the attribute  `final` to the RealOutput. For example, in MSL 3.2, TemperatureSensor declares
```
Modelica.Blocks.Interfaces.RealOutput T annotation (Placement(
        transformation(extent={{90,-10},{110,10}}, rotation=0)));
```
which was changed in 3.2.1 beta to
```
  Modelica.Blocks.Interfaces.RealOutput T(final quantity="ThermodynamicTemperature", final unit="K") 
    "Absolute temperature as output signal"                                                                                                  annotation (Placement(transformation(extent={{90,-10},{110,10}}, rotation=0)));
```
This change is not backward compatible. For example, if a model instantiates
```
Modelica.Thermal.HeatTransfer.Sensors.TemperatureSensor temSen(
    T(final quantity="Temperature",
      final unit = "K", displayUnit = "degC", min=0)) 
    "Temperature sensor of metal";
```
then the model is not valid, and Dymola will stop the translation due to overriding final modifier.

A work-around is to remove the `final` attribute in the next beta release of 3.2.1

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1125

**Comment by modelica-trac-importer** [comment]
**Comment by ahaumer on 25 May 2013 08:00 UTC**
fixed by eab8aa36196eecc8573bc3721f376165efb35b9c

**Comment by modelica-trac-importer** [comment]
**Comment by mwetter on 25 May 2013 14:30 UTC**
The same problem exist for `HeatFlowSensor` and `ConditionalFixedHeatFlowSensor`.

**Comment by modelica-trac-importer** [comment]
**Comment by mwetter on 25 May 2013 14:32 UTC**
Fixed in 8575d25c9d9f6b7f93f41fd8cb5d78e5c79e947a.

**Comment by modelica-trac-importer** [comment]
**Comment by mwetter on 3 Jul 2013 22:42 UTC**
It turns out that removing the `final`  attribute is not sufficient to ensure backward compatibility. The declaration of the `quantity` must be removed for models to be backward compatible. The attached file is an example that works with MSL 3.2, but not with 3.2.1 beta 3 (build 0), using the trunk eec12dc4693f283ed397cba1ae86a7494488a369.

In this case, a library used models that declared

```

Modelica.Blocks.Interfaces.RealOutput T1(quantity="Temperature");

```
When connected to a temperature sensor, the model fails to translate in Dymola 2014 because the quantities are different when MSL 3.2.1 is used. In MSL 3.2, the model works because `Modelica.Thermal.HeatTransfer.Sensors.TemperatureSensor` does not declare a quantity.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 4 Jul 2013 06:51 UTC**
Sensor.mo works in SimulationX in both cases:
 * Using MSL 3.2: T1=293.15 [-], temperatureSensor.T=293.15 [-]
 * Using MSL 3.2.1: T1=293.15 [-], temperatureSensor.T=293.15 [K]

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Jul 2013 07:51 UTC**
Replying to [comment:4 mwetter]:
> It turns out that removing the `final`  attribute is not sufficient to ensure backward compatibility. The declaration of the `quantity` must be removed for models to be backward compatible. The attached file is an example that works with MSL 3.2, but not with 3.2.1 beta 3 (build 0), using the trunk eec12dc4693f283ed397cba1ae86a7494488a369.
> 
> In this case, a library used models that declared
> 
> ```
> 
> Modelica.Blocks.Interfaces.RealOutput T1(quantity="Temperature");
> 
> ```
> When connected to a temperature sensor, the model fails to translate in Dymola 2014 because the quantities are different when MSL 3.2.1 is used. In MSL 3.2, the model works because `Modelica.Thermal.HeatTransfer.Sensors.TemperatureSensor` does not declare a quantity.

I understand: Of course, there is a small bug in the user model, because it uses the wrong quantity (should use "ThermodynamicTemperature" instead of "Temperature"). However, such tiny errors can easily occur in user models, and then it is annoying that the model does no longer work with a new version of MSL. Therefore, I agree to remove all quantity information from sensors and will do it just now.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 4 Jul 2013 08:07 UTC**
I wonder what the final solution for this would be then. I mean we can not make sure that the new version of the MSL will work for incorrect user models. All we can assure is that it will work with **correct** models. How are we then ever going to fix the sensor outputs, since I don't see how you could correct an incorrect user model with a conversion script in the future. Basically this means we are never going to have sensor outputs with quantities right?

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 4 Jul 2013 08:09 UTC**
The question for me then is if SimulationX keeps its own MSL 3.2.1 (as units in sensors are far more user-friendly).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Jul 2013 08:17 UTC**
Replying to [comment:7 dietmarw]:


> I wonder what the final solution for this would be then. I mean we can not make sure that the new version of the MSL will work for incorrect user models. All we can assure is that it will work with **correct** models. How are we then ever going to fix the sensor outputs, since I don't see how you could correct an incorrect user model with a conversion script in the future. Basically this means we are never going to have sensor outputs with quantities right?
>

Yes, also in the future, quantities will not be used in sensors. This is not a big issue. The most important point is that the units are the same. There are anyway only a few physical quantities where the quantity is different and the 7 base units are the same.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Jul 2013 08:24 UTC**
Replying to [comment:8 beutlich]:
> The question for me then is if SimulationX keeps its own MSL 3.2.1 (as units in sensors are far more user-friendly).

If SimulationX has a better concept for units, it would be best to propose this as new feature for Modelica by providing a Modelica Change Proposal (https://svn.modelica.org/projects/MCP-Ideas).

My intention was to improve units for the next Modelica Spec. by using the FMI 2.0 definition of units, in order to provide the missing definition of MSL units with respect to the 7 SI base units (so use the xml-schema definition of FMI 2.0 to define units and relate this definition "somehow" in Modelica.SIunits). If there are also good ideas to make unit handling more convenient in a model, then this is the right time to propose it.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 4 Jul 2013 08:26 UTC**
So we are going to stop features in the MSL in favour of incorrect user models? Did I understand this correctly? That's a strange strategy really and as comment:8 pointed out will lead to diversion from the source MSL which would be a shame.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 4 Jul 2013 09:16 UTC**
The user model Sensor.mo really is wrong as "Temperature" is not used as quantity at all in the MSL. In this case the preferred solution is to fix the wrong model.

Other quick suggestions:
 * Use MSL 3.2 with Dymola 2014.
 * Relax quantity check in newer Dymola from error to warning only.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Jul 2013 09:47 UTC**
I disussed with ITI (Gerd Kurzbach, Thomas Beutlich) on the phone. The issue is more complex than I anticipated. The conclusion was to wait and discuss this on the next web meeting for the ModelicaSpec 3.2 rev.2 (most likely next Tuesday). Here are the issues:

**Modelica Specification**:
The quantity and unit attributes are defined, but not the semantics of it. Only the syntax of units is defined and that the unit attribute "is used in equations". So, the specification is ambiguous. Especially, the meaning of "quantity" is not defined and it is not defined in which way variables can be combined, e.g., in connection equations. It is also not defined that attributes min, max, start have all the same unit (although this seems obvious).

When the unit handling was designed in the 90tes, the intention was that only variables can be connected in a connection, if they have the same "quantity" name and the units can be transformed to the same base units. However, this seems to be not stated in the specification.

**Dymola**
Current Dymola gives an error if in a connection statement two or more different quantity names are used. If one variable has a quantity and the other does not have a quantity name, then this is accepted. Dymola deduces the unit-menu for unit conversion from the SI unit: E.g. "rad/s" is the SI unit, and then "deg/s", "rpm" etc. are defined with respect to "rad/s".

**SimulationX**
SimulationX does not take into account the quantity name in a connection check (so SimulationX allows to connect two variables with different quantity names). SimulationX deduces the unit-menu from the quantity name. Several names can be used defining the same quantity (e.g. ThermodynamicTemperature, Temperature, CelsiusTemperature define the same physical quantity). So removing the quantity name from sensors, will remove also the unit-menu for sensors.

**MSL**
The quantity check is essential for Media/Fluid components, since the Media name is used as quantity name, e.g., for MassFlowRate. Therefore, Dymola gives an error, if components are connected and different Media are defined in the connectors of these components (this situation can easily occur). With current Modelica and MSL this is essential. However, this is a hack and would not be needed, if a medium could be defined at one place and then it is automatically propagated through connections.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Jul 2013 09:50 UTC**
Sorry forgot one issue:
The usage of quantity names is also not well defined. There is the convention to use the quantity names from the MSL, but it is not stated that one has to. So, if a user uses different quantity names in a model, one cannot state that this is an error in a user model.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 4 Jul 2013 10:32 UTC**
Replying to [comment:14 otter]:
> So, if a user uses different quantity names in a model, one cannot state that this is an error in a user model.

Exactly! That's the reason why Dymola must not raise an error but rather should warn.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Jul 2013 12:38 UTC**
Replying to [comment:15 beutlich]:
> Replying to [comment:14 otter]:
> > So, if a user uses different quantity names in a model, one cannot state that this is an error in a user model.
> Exactly! That's the reason why Dymola must not raise an error but rather should warn. 

However, we also need a (different) mechanism for ensuring that you get errors when connect different media in the fluid library. 

I don't see that we can find such a solution for MSL 3.2.1 and Spec 3.2rev2, but we clearly need a better solution for the future.

In addition the quantities were added in a clumsy way for MSL 3.2.1 and I would propose delay adding them until we have a better mechanics. The problem is that since we cannot 
reuse the types we have to manually add the text where appropriate, which increases the risk of errors (both in user libraries and MSL) - and we should instead design a better alternative where we reuse Modelica.SIunits in the sensors (I believe there is already a ticket about this).

E.g. the new definition caused a problem since (at least in some build),
Modelica.Electrical.Analog.Sensors.VoltageSensor had quantity="PotentialDifference"
whereas   
```
  type ElectricPotential = Real (final quantity="ElectricPotential", final unit=
         "V");
  type Voltage = ElectricPotential;
```

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 10 Jul 2013 14:07 UTC**
Discussion and decision at web/phone meeting:

**Issue 1:**
Inputs and outputs of physical models got "final" definitions. However, in user models quantity and unit definitions might already been present and with MSL 3.2.1 an error will occur, since a final modifier is overwritten

Decision:
Remove "final" modifiers from the newly introduced input/output signals.
(the quantities/units that have been newly introduced are listed here:
https://trac.modelica.org/Modelica/attachment/ticket/876/T876.diff)

**Issue 2: **
Semantics of quantity is not defined in specification and e.g. Modelica.Fluid relies on a specific semantics (not allowed to connect different quantities).

Decision:
Do not define the semantics of quantity in Modelica Spec. 3.2 rev2, since tool vendors use it already with different semantics.
("quantities" must be identical when connecting in some tool, or can be different in another tool).
This should be fixed if a complete solution for units/quantities is developed (but there is not enough time to do this for rev.2).

**Issue 3:**
Quantities are added in an error prone way in MSL 3.2.1 since quantity and unit definitions are explicitely given and the underlying type is not used.

Discussion:
The goal is to introduce a new language element to make this less error prone. However, even if a new language element would be introduced, such as:

  RealOutput y (redeclare Modelica.SIunits.Temperature).

this would not help for MSL, since Modelica.SIunits has "final" modifiers and therefore such a change can only be made if at the same time a conversion script removes corresponding modifiers from user models.

**Issue 4:**
Sensors get “quantity” definitions in MSL 3.2.1, and user models might use different quantity definitions (and in one case, two different quantity definitions are used in MSL 3.2.1). Some tools, like Dymola, therefore give an error in such a situation. SimulationX uses the quantity definition to provide
a unit-menu in order to display the variable.

Decision:
Remove the newly introduced quantity definitions for inputs/outputs.
(SimulationX plans to include a new feature to deduce the unit menu also
from a unit if no quantity is present).

Thomas (Beutlich) will perform the changes in MSL. Example:
   RealOutput y(unit="K");

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 11 Jul 2013 10:14 UTC**
Resolved by 9256f7bba39321d331f2a2ba5d8ee2e1bb0b47f6.
